File: perl/pe007.pl
#!/usr/bin/perl -w
use strict;

my $last = 10001;
my @primes = qw(2 3);

my $test = $primes[-1] + 2;
while (@primes < $last) {
  if (is_prime($test, \@primes)) {
    push @primes, $test;
  }
  $test += 2;
}
print $primes[-1],"\n";

sub is_prime {
  my $test = shift;
  my $known_primes_ref = shift;

  my $root = sqrt($test);
  foreach my $number (@$known_primes_ref) {
    return 1 if ($number > $root);
    return 0 unless ($test % $number);
  }
  return 1;
}

File: ruby/pe007.rb
#!/usr/bin/ruby

def is_prime(test, primes)
  root = Math.sqrt(test)

  primes.each do |i|
    if i > root
      return true
    end
    if test % i == 0
      return false
    end
  end
end

primenum = 10001
primes = Array[2, 3]

test = 5
while (primes.length < primenum)
  if (is_prime(test, primes))
    primes.push(test)
  end
  test += 2
end

puts(primes[-1])
